controllers/maascluster_controller.go

```diff
@@ -168,48 +168,50 @@ func (r *MaasClusterReconciler) reconcileDNSAttachments(clusterScope *scope.Clus
 		return errors.Wrapf(err, "unable to find preferred subnets")
 	}
 
+	// Build desired IPs first (no MAAS call)
 	var runningIpAddresses []string
-
-	currentIPs, err := dnssvc.GetAPIServerDNSRecords()
-	if err != nil {
-		return errors.Wrap(err, "Unable to get the dns resources")
-	}
-
-	machinesPendingAttachment := make([]*infrav1beta1.MaasMachine, 0)
-	machinesPendingDetachment := make([]*infrav1beta1.MaasMachine, 0)
-
 	for _, m := range machines {
 		if !IsControlPlaneMachine(m) {
 			continue
 		}
-
 		machineIP := getExternalMachineIP(clusterScope.Logger, preferredSubnets, m)
-		attached := currentIPs.Has(machineIP)
 		isRunningHealthy := IsRunning(m)
-
 		if !m.DeletionTimestamp.IsZero() || !isRunningHealthy {
-			if attached {
-				clusterScope.Info("Cleaning up IP on unhealthy machine", "machine", m.Name)
-				machinesPendingDetachment = append(machinesPendingDetachment, m)
-			}
-		} else if IsRunning(m) {
-			if !attached {
-				clusterScope.Info("Healthy machine without DNS attachment; attaching.", "machine", m.Name)
-				machinesPendingAttachment = append(machinesPendingAttachment, m)
-			}
-
+			continue
+		}
+		if machineIP != "" {
 			runningIpAddresses = append(runningIpAddresses, machineIP)
 		}
-		//r.Recorder.Eventf(machineScope.MaasMachine, corev1.EventTypeNormal, "SuccessfulDetachControlPlaneDNS",
-		//	"Control plane instance %q is de-registered from load balancer", i.ID)
-		//runningIpAddresses = append(runningIpAddresses, m.)
 	}
 
-	if err := dnssvc.UpdateDNSAttachments(runningIpAddresses); err != nil {
+	// Early-exit gate using last-applied hash
+	const lastAppliedAnn = "infrastructure.cluster.x-k8s.io/last-applied-dns-hash"
```

Comment:
For clarity and reuse, can we add this constant a better place, like at the top of the file instead of inside the function?